File: oop/lab2/java/PascalTriangleRow.java
import java.util.ArrayList;
import java.util.List;

public class PascalTriangleRow {
    private final List<Integer> row;

    public PascalTriangleRow(int n) throws IllegalArgumentException {
        if (n < 0) {
            throw new IllegalArgumentException("Invalid argument: " + n);
        }
        try {
            this.row = Binomials.generateRow(n);
        } catch (Exception e) {
            throw new IllegalArgumentException(e.getMessage());
        }
    }

    public int getCoefficient(int m) throws IllegalArgumentException, IndexOutOfBoundsException {
        if (m < 0) {
            throw new IllegalArgumentException("Index must be greater or equal to 0. Got " + m + " instead.");
        }
        if (m >= this.row.size()) {
            throw new IndexOutOfBoundsException("Range exceeded. Index must be less than or equal to "
                    + (this.row.size() - 1) + ". Got " + m + " instead.");
        }

        return this.row.get(m);
    }
}

File: 2024_semester_2/oop/lab2/cpp/src/pascalTriangleRow.cpp
#include <stdexcept>
#include <string>

#include <binomials.hpp>

#include <pascalTriangleRow.hpp>

PascalTriangleRow::PascalTriangleRow(const int n) noexcept(false)
{
    if (n < 0)
    {
        throw std::invalid_argument(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "Invalid argument: " + std::to_string(n));
    }
    this->row = Binomials::generateRow(n);
}

int PascalTriangleRow::getCoefficient(int m) noexcept(false)
{
    if (m < 0)
    {
        throw std::invalid_argument(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "index must be greater or equal to 0. Got " + std::to_string(m) + " instead.");
    }
    if (static_cast<size_t>(m) >= this->row.size())
    {
        throw std::out_of_range(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "range exceeded. Index must be less than or equal to " + std::to_string(this->row.size() - 1) + ". Got " + std::to_string(m) + " instead.");
    }

    return this->row[static_cast<size_t>(m)];
}
